### PR TITLE
Required IDs and more informative error messages

### PR DESCRIPTION
**Pull request by dmdashenkov**
In this PR we add support for the required ID rules.

When a special kind of message is defined, an event, a command, a rejection, or an entity state, its first field must be the ID field. The ID fields are required by default (unless `(required) = false`).

Also, in this PR we tighten the rule for required collections. Previously, we only checked the collection to not be empty. Now, we also check the elements to be non-empty (non-default). This does not apply to primitive types that have no concept of an empty value, such as numbers and `bool`s.

### CODE REVIEW

File: model/src/main/java/io/spine/validation/RequiredIdPatternPolicy.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+import static io.spine.validation.SourceFiles.findFirstField;
+
+/**
+ * A policy that marks ID fields in entity state messages and signal messages as required.
+ *
+ * <p>The messages are discovered via the file patterns, specified in {@link ValidationConfig}.
+ * If ProtoData runs with no config, this policy never produces any validation rules.
+ */
+final class RequiredIdPatternPolicy extends RequiredIdPolicy {
```
Comment (armiol) [inline model/src/main/java/io/spine/validation/RequiredIdPatternPolicy.java:48]: Let's describe the difference between this policy and `RequiredIdOptionPolicy`. As discussed, we need to guide the potential users on which to apply and when.

File: model/src/main/proto/spine/validation/config.proto [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+option java_package = "io.spine.validation";
+option java_outer_classname = "ConfigProto";
+option java_multiple_files = true;
+
+import "spine/tools/mc/java/config.proto";
+
+// The container type for the validation config.
+//
+// ProtoData supports multiple configurations merges into a single object. In order for
+// the Validation library to be satisfied with the config, it mast always have
```
Comment (armiol) [inline model/src/main/proto/spine/validation/config.proto:17]: `must`.

**Review by armiol — CHANGES_REQUESTED** [review]
@dmdashenkov LGTM in general. Please see my comments.

File: model/src/main/java/io/spine/validation/RequiredIdPatternPolicy.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+                                    .anyMatch(pattern -> matches(file, pattern));
+        if (!match) {
+            return withNothing();
+        }
+        TypeName type = event.getType();
+        Field field = findFirstField(type, file, this);
+        return withField(field);
+    }
+
+    private static ImmutableList<FilePattern> allPatterns(MessageMakers markers) {
```
Comment (alexander-yevsyukov) [inline model/src/main/java/io/spine/validation/RequiredIdPatternPolicy.java:70]: Please consider transforming this conversion to Kotlin extension for `MessageMakers`.

File: model/src/main/java/io/spine/validation/ValidationPolicy.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+/**
+ * A policy that reacts to an event with a {@link RuleAdded} event.
+ *
+ * <p>May ignore an event and return {@code Nothing} if necessary.
+ *
+ * @param <E>
+ *         the type of the event to react to
+ */
+public abstract class ValidationPolicy<E extends EventMessage>
+        extends Policy<E> {
```
Comment (alexander-yevsyukov) [inline model/src/main/java/io/spine/validation/ValidationPolicy.java:45]: Please put `extends` on the previous line. It should fit.

File: model/src/main/java/io/spine/validation/Rules.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    /**
+     * Prevents the utility class instantiation.
+     */
+    private Rules() {
+    }
+
+    /**
+     * Converts the given {@code rule} to an event.
+     */
+    static RuleAdded toEvent(Rule rule, TypeName type) {
```
Comment (alexander-yevsyukov) [inline model/src/main/java/io/spine/validation/Rules.java:47]: Please consider transforming this into a Kotlin extension for the `Rule` type.

Please also document the `type` parameter.

File: model/src/main/java/io/spine/validation/RequiredRule.java [hunk 6]
```diff
@@ -48,32 +53,89 @@ private RequiredRule() {
...
+    }
+
+    private static SimpleRule rule(Field field,
+                                   Value value,
+                                   String errorMessage,
+                                   String defaultErrorMessage,
+                                   boolean distibute) {
+        String msg = errorMessage.isEmpty() ? defaultErrorMessage : errorMessage;
+        return SimpleRule
                 .newBuilder()
```
Comment (alexander-yevsyukov) [inline model/src/main/java/io/spine/validation/RequiredRule.java:102]: Please put `newBuilder()` on the previous line.

File: model/src/main/java/io/spine/validation/RequiredRule.java [hunk 7]
```diff
@@ -48,32 +53,89 @@ private RequiredRule() {
...
-                "Field `%s.%s` of type `%s` does not support `(required)` validation.",
-                typeUrl, fieldName, type
-        ));
+    private static Rule wrap(SimpleRule r) {
+        return Rule.newBuilder()
+                .setSimple(r)
+                .build();
+    }
+
+    private static Rule wrap(CompositeRule r) {
```
Comment (alexander-yevsyukov) [inline model/src/main/java/io/spine/validation/RequiredRule.java:117]: Can these two methods be a part of a mixin interface common for `SimpleRule` and `CompositeRule`?

File: model/src/main/java/io/spine/validation/RequiredPolicy.java [hunk 8]
```diff
@@ -65,15 +67,24 @@
...
+        return Rules.toEvent(rule, declaration.getDeclaringType());
+    }
+
+    private static IllegalStateException doesNotSupportRequired(Field field) {
+        String fieldName = field.getName()
+                                .getValue();
+        String typeUrl = typeUrl(field.getDeclaringType());
+        PrimitiveType type = field.getType()
+                                  .getPrimitive();
+        return new IllegalStateException(format(
```
Comment (alexander-yevsyukov) [inline model/src/main/java/io/spine/validation/RequiredPolicy.java:85]: We have `Exceptions.newIllegalStateException()` which does exactly this. Why not use it?

**Review by alexander-yevsyukov — APPROVED** [review]
LGTM, with a few suggestions to consider before merging.

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

**Review by armiol — APPROVED** [review]

**Pull request merged**